core/src/main/java/com/linecorp/armeria/client/ConnectionPoolMetrics.java

```diff
@@ -15,50 +15,129 @@
+
     private final MeterRegistry meterRegistry;
     private final MeterIdPrefix idPrefix;
+    @GuardedBy("lock")
+    private final Map<List<Tag>, Meters> metersMap = new HashMap<>();
+    private final ReentrantShortLock lock = new ReentrantShortLock();
 
     /**
      * Creates a new instance with the specified {@link Meter} name.
      */
     ConnectionPoolMetrics(MeterRegistry meterRegistry, MeterIdPrefix idPrefix) {
-        this.idPrefix = requireNonNull(idPrefix, "idPrefix");
-        this.meterRegistry = requireNonNull(meterRegistry, "meterRegistry");
+        this.idPrefix = idPrefix;
+        this.meterRegistry = meterRegistry;
     }
 
     void increaseConnOpened(SessionProtocol protocol, InetSocketAddress remoteAddr,
                             InetSocketAddress localAddr) {
-        meterRegistry.counter(idPrefix.name(),
-                              idPrefix.tags(PROTOCOL, protocol.name(),
-                                            REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
-                                            LOCAL_IP, localAddr.getAddress().getHostAddress(),
-                                            STATE, "opened"))
-                     .increment();
+        final List<Tag> commonTags = commonTags(protocol, remoteAddr, localAddr);
+        lock.lock();
+        try {
+            final Meters meters = metersMap.computeIfAbsent(commonTags,
+                                                            key -> new Meters(idPrefix, key, meterRegistry));
+            meters.increment();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private List<Tag> commonTags(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                 InetSocketAddress localAddr) {
+        return idPrefix.tags(PROTOCOL, protocol.name(),
+                             REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
+                             LOCAL_IP, localAddr.getAddress().getHostAddress());
     }
 
     void increaseConnClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
                             InetSocketAddress localAddr) {
-        meterRegistry.counter(idPrefix.name(),
-                              idPrefix.tags(
-                                      PROTOCOL, protocol.name(),
-                                      REMOTE_IP, remoteAddr.getAddress().getHostAddress(),
-                                      LOCAL_IP, localAddr.getAddress().getHostAddress(),
-                                      STATE, "closed"))
-                     .increment();
+
+        final List<Tag> commonTags = commonTags(protocol, remoteAddr, localAddr);
+        lock.lock();
+        try {
+            final Meters meters = metersMap.get(commonTags);
+            if (meters != null) {
+                meters.decrement();
+                assert meters.activeConnections() >= 0 : "active connections should not be negative" + meters;
```

Comment:
Question) Will the `AssertionError` be caught anywhere?

```suggestion
                checkState(meters.activeConnections() >= 0, : "active connections should not be negative" + meters);
```

Reply:
The `AssertionError` should be logged by:
https://github.com/line/armeria/blob/9823b46fce5693810539e26dfae7083e24f57586/core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java#L516-L523
I believe `assert` would be proper because if the value becomes negative, it is a bug but not a bad state.

Reply:
I see, I thought `Error` didn't inherit from `Exception`. No problem if it is somehow logged and propagated to the user

Reply:
Good point. I don't see we need to log only `Exception`s. Let me change it to `Throwable`.